### ISSUE TITLE
`beforeNavigate` fires when the page is reloaded

### ISSUE THREAD

**Issue by aradalvand** [sveltejs/kit#4383]
### Describe the bug

Previously, I was using the `sveltekit:navigation-start` and `sveltekit:navigation-end` events that were fired on `window` to show/hide a navigation loader on the page.

Now, it seems like the aforementioned events [were removed](https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#100-next285) in favor of the `beforeNavigate`/`afterNavigate` lifecycle functions.

However, I expected the `beforeNavigate` lifecycle event to get fired only when client-side routing is triggered, just like how the good old `sveltekit:navigation-start` used to work, but I noticed that it gets fired even when you refresh the page, or manually modify the URL in the browser address bar and hit enter to leave the page.
https://github.com/sveltejs/kit/blob/e28bc390e033e38585fd1640da65c22f973afd2c/packages/kit/src/runtime/client/router.js#L110-L119
I don't understand the reason behind this, if someone wants to listen for `beforeunload` they could do so by simply registering a listener for that event; correct me if I'm wrong but there's no need for SvelteKit to additionally fire `beforeNavigate`, as that can be completely undesirable in certain cases, such as in the case of a client-side navigation loader, and it's also not consistent with the behavior of the old `sveltekit:navigation-start` event.
My primary suggestion is that the behavior should change to match that of the `sveltekit:navigation-start` event, namely that it should only be fired when client-side navigation occurs. This is the most expected behavior.
But at the very least, I think the callback function should be passed a boolean argument that indicates whether the navigation was triggered by the client-side router or not:
```js
beforeNavigate(({ client }) => {
    if (client) {
      // Blah blah blah...
    }
});
```

It's also worth noting that the `$navigating` store works as expected in this regard, namely that it gets populated only in the case of client-side routing. So, the current behavior of `beforeNavigate` is also inconsistent with the `$navigating` store.

### System Info

```shell
System:
    OS: Windows 10 10.0.19044
    CPU: (4) x64 Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
    Memory: 885.61 MB / 7.89 GB
  Binaries:
    Node: 14.15.0 - C:\Program Files\nodejs\node.EXE     
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD       
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD        
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (99.0.1150.39)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @sveltejs/kit: ^1.0.0-next.301 => 1.0.0-next.301
    svelte: ^3.46.4 => 3.46.4
```

### Severity

serious, but I can work around it

**Comment by AradAral** [comment]
Actually I also just noticed that `afterNavigate` gets fired when the page is loaded too. This can be undesirable as well. I think the API here needs to be rethought a little bit.

**Comment by Rich-Harris** [comment]
It's very much intentional that `beforeNavigate` fires in `beforeunload` — if the user is on a page where they're editing some data, they want to be prevented from losing unsaved changes. They don't care if they're about to lose unsaved changes because they clicked on an internal link or an external one.

That said, the current API isn't great. Firstly, `to` should be populated in both cases (right now, you can distinguish between client-side and native navigations by whether or not `to` is `null`, but that's weird). Secondly, `beforeNavigate` should fire when you navigate away from the page even if client-side routing is disabled, I think.

Finally, as you say, there should be an explicitly property that says whether this is a client-side navigation or a native one. A `client` boolean is one option, but with a less ambiguous name. Another option would to consider this alongside other information we might want in `beforeNavigate` that I haven't got round to raising an issue for — you might want to know whether a navigation was triggered by a link click or a `goto` or a `popstate`, for example:

```js
beforeNavigate(({ from, to, type, cancel }) => {
  if (type === 'back') { // or maybe 'popstate' with a separate property indicating the delta
    animation_direction = -1;
  }

  if (type === 'forward') {
    animation_direction = 1;
  }

  if (type === 'reload') {
    // we're in beforeunload
  }
});
```

To your second comment, it's also intentional that `afterNavigate` fires on page load, aka _after_ you _navigated_ to the page. `from` will be `null` in that case.

**Comment by AradAral** [comment]
> It's very much intentional that beforeNavigate fires in beforeunload — if the user is on a page where they're editing some data, they want to be prevented from losing unsaved changes. They don't care if they're about to lose unsaved changes because they clicked on an internal link or an external one.

Oh okay, that's right, makes sense.

> That said, the current API isn't great. Firstly, to should be populated in both cases (right now, you can distinguish between client-side and native navigations by whether or not to is null, but that's weird).

Totally agree.

> Finally, as you say, there should be an explicitly property that says whether this is a client-side navigation or a native one. A client boolean is one option, but with a less ambiguous name. Another option would to consider this alongside other information we might want in beforeNavigate that I haven't got round to raising an issue for — you might want to know whether a navigation was triggered by a link click or a goto or a popstate, for example:

The `type` property makes sense too — although `reload` should be called `unload` since `beforeunload` could fire as a result of not just reloading the page but any kind of native browser navigation — and its counterpart in `afterNavigate` should presumably be called `load`, I guess.

The only thing I'm a bit confused about is whether or not all `type`s other than `unload`/`load` can be assumed to all be client-side navigation (provided, of course, that client-side routing is enabled)? So if you want to, for example, emulate the behavior of the old `sveltekit:navigation-start` in `beforeNavigate`, you could just check whether the `type` is `unload` or not?

In other words, with a boolean `client` property, this:
```js
beforeNavigate(({ client }) => {
    if (!client) return;
});
```
would be functionally equivalent to the following with a `type` property:
```js
beforeNavigate(({ type }) => {
    if (type == 'unload') return;
});
```
Am I right?

**Comment by Rich-Harris** [comment]
Yep, `type === 'unload'` and `!client` would be equivalent. All other types would be client-side navigations

**Comment by AradAral** [comment]
Well great, then I think the `type` property is the way to go.

**Comment by dummdidumm** [comment]
Closed via #6537
